src/main/java/com/gregtechceu/gtceu/api/recipe/ingredient/IntProviderFluidIngredient.java

```diff
@@ -0,0 +1,343 @@
+    }
+
+    public static IntProviderFluidIngredient of(FluidStack stack, IntProvider provider) {
+        return IntProviderFluidIngredient.of(FluidIngredient.of(stack), provider);
+    }
+    public static IntProviderFluidIngredient of(FluidStack stack, int min, int max) {
+        return IntProviderFluidIngredient.of(FluidIngredient.of(stack), UniformInt.of(min, max));
+    }
+
+    public static IntProviderFluidIngredient of(IntProvider countProvider, Fluid... fluids) {
+        return IntProviderFluidIngredient.of(Arrays.stream(fluids), countProvider, null);
+    }
+
+    private static IntProviderFluidIngredient of(Stream<Fluid> stacks, IntProvider countProvider, CompoundTag nbt) {
+        return IntProviderFluidIngredient.fromValues(
+                stacks.filter(stack -> stack != null && !stack.isSame(Fluids.EMPTY)).map(FluidValue::new), 1000, nbt,
+                countProvider);
+    }
+//
+//    public static IntProviderFluidIngredient of(IntProvider countProvider, FluidStack... stacks) {
+//        return IntProviderFluidIngredient.of(Arrays.stream(stacks).map(FluidStack::getFluid),
+//                countProvider, stacks.length == 0 ? null : stacks[0].getTag());
+//    }
+//
+//    public static IntProviderFluidIngredient of(Stream<Fluid> stacks,  CompoundTag nbt) {
+//        return IntProviderFluidIngredient.fromValues(
+//                stacks.filter(stack -> stack != null && !stack.isSame(Fluids.EMPTY)).map(FluidValue::new), 1000, nbt,
+//                null);
+//    }
+
+    public static IntProviderFluidIngredient of(TagKey<Fluid> tag, IntProvider countProvider) {
+        return IntProviderFluidIngredient.fromValues(Stream.of(new IntProviderFluidIngredient.TagValue(tag)), 1000, null, countProvider);
+    }
+
+    public static IntProviderFluidIngredient of(TagKey<Fluid> tag, int min, int max) {
+        return IntProviderFluidIngredient.fromValues(Stream.of(new IntProviderFluidIngredient.TagValue(tag)), 1000, null, UniformInt.of(min, max));
+    }
+
+    public static IntProviderFluidIngredient of(TagKey<Fluid> tag, IntProvider countProvider, CompoundTag nbt) {
+        return IntProviderFluidIngredient.fromValues(Stream.of(new IntProviderFluidIngredient.TagValue(tag)), 1000, nbt, countProvider);
+    }
+
+    public static IntProviderFluidIngredient of(TagKey<Fluid> tag, int min, int max, CompoundTag nbt) {
+        return IntProviderFluidIngredient.fromValues(Stream.of(new IntProviderFluidIngredient.TagValue(tag)), 1000, nbt, UniformInt.of(min, max));
+    }
+
+    public CompoundTag writeToNBT(CompoundTag nbt) {
+        nbt.putString("FluidName", ForgeRegistries.FLUIDS.getKey(this.inner.getStacks()[0].getFluid()).toString());
+        nbt.putInt("Amount", this.amount);
+        if (this.nbt != null) {
+            nbt.put("Tag", this.nbt);
+        }
+        nbt.putInt("Minimum", this.getCountProvider().getMinValue());
+        nbt.putInt("Maximum", this.getCountProvider().getMaxValue());
```

Comment:
don't do it like this, encode the actual int provider instead.
```suggestion
        nbt.put("count_provider", IntProvider.CODEC.encodeStart(NbtOps.INSTANCE, countProvider)
                .getOrThrow(false, GTCEu.LOGGER::error));
```

Reply:
also, the nbt keys you have are different from the JSON keys. make them the same